File: examples/workflows/shared.mjs
import {readFile} from 'node:fs/promises';
import {fileURLToPath} from 'node:url';
import {createHash} from 'node:crypto';
import {AgentTrunkClient} from '@agenttrunk/sdk';
export function substitute(value,env=process.env){
 if(typeof value==='string')return value.replace(/\$\{(AGENTTRUNK_[A-Z_]+)\}/g,(_,key)=>{if(!env[key])throw Error(`Missing ${key}`);return env[key];});
 if(Array.isArray(value))return value.map(v=>substitute(v,env));
 if(value&&typeof value==='object')return Object.fromEntries(Object.entries(value).map(([k,v])=>[k,substitute(v,env)]));return value;
}
export async function run(url,operation,execute){
 if(process.argv[1]!==fileURLToPath(url))return;
 try{
  const flags=process.argv.slice(2);if(flags.some(f=>!['--execute','--yes'].includes(f)))throw Error('Invalid flags');
  const raw=await readFile(new URL('./request.json',url),'utf8');
  const mutation=['contexts.publish','releases.open'].includes(operation);
  if(!flags.includes('--execute')){console.log(JSON.stringify({operation,effect:mutation?'write':'read',inputSha256:createHash('sha256').update(raw).digest('hex'),requiredEnvironment:[...new Set(raw.match(/AGENTTRUNK_[A-Z_]+/g)??[])],preview:true}));return;}
  if(mutation&&!flags.includes('--yes'))throw Error('Write approval required');
  if(!process.env.AGENTTRUNK_ACCESS_TOKEN)throw Error('Missing runtime token');
  const client=new AgentTrunkClient({accessToken:()=>process.env.AGENTTRUNK_ACCESS_TOKEN,baseUrl:process.env.AGENTTRUNK_API_URL,maxRetries:0});
  console.log(JSON.stringify(await execute(client,substitute(JSON.parse(raw))),null,2));
 }catch{console.error('Workflow failed. Check configuration and permissions; reconcile an uncertain write before retrying.');process.exitCode=1;}
}

File: examples/workflows/shared.py
import hashlib
import json
import os
import re
import sys
from agenttrunk import AgentTrunk

def kwargs(value):
    return {re.sub(r'[A-Z]', lambda m: '_' + m[0].lower(), key): val for key, val in value.items()}

def substitute(value):
    if isinstance(value, str):
        return re.sub(r'\$\{(AGENTTRUNK_[A-Z_]+)\}', lambda m: os.environ[m[1]], value)
    if isinstance(value, list): return [substitute(v) for v in value]
    if isinstance(value, dict): return {k: substitute(v) for k, v in value.items()}
    return value

def run(path, operation, execute):
    try:
        flags = sys.argv[1:]
        if any(flag not in ('--execute', '--yes') for flag in flags): raise ValueError('Invalid flags')
        raw = (path.parent / 'request.json').read_bytes()
        mutation = operation in ('contexts.publish', 'releases.open')
        if '--execute' not in flags:
            print(json.dumps(dict(operation=operation, preview=True, effect='write' if mutation else 'read', inputSha256=hashlib.sha256(raw).hexdigest(), requiredEnvironment=sorted(set(re.findall(r'AGENTTRUNK_[A-Z_]+', raw.decode()))))))
            return
        if mutation and '--yes' not in flags: raise ValueError('Write approval required')
        os.environ['AGENTTRUNK_ACCESS_TOKEN']
        client = AgentTrunk(access_token=lambda: os.environ['AGENTTRUNK_ACCESS_TOKEN'], base_url=os.getenv('AGENTTRUNK_API_URL', 'https://api.agenttrunk.ai'), max_retries=0)
        result = execute(client, substitute(json.loads(raw)))
        if hasattr(result, 'model_dump_json'): print(result.model_dump_json(by_alias=True))
        else: print(json.dumps(result))
    except Exception:
        print('Workflow failed. Check configuration and permissions; reconcile an uncertain write before retrying.', file=sys.stderr)
        sys.exit(1)
